scripts/add_electricity.py

```diff
@@ -318,45 +348,55 @@ def attach_storageunits(n, costs, technologies, extendable_carriers):
     lookup_dispatch = {"H2": "fuel cell", "battery": "battery inverter"}
 
-    buses_i = n.buses.index
+    number_microgrids = len(number_microgrids.keys())
+    microgrid_ids = [f"microgrid_{i+1}" for i in range(number_microgrids)]
 
-    # Iterate through each storage technology
-    for tech in technologies:
-        # Add the storage units to the power network
-        n.madd(
-            "StorageUnit",
-            buses_i,
-            " " + tech,
-            bus=buses_i,
-            carrier=tech,
-            p_nom_extendable=True,
-            capital_cost=costs.at[tech, "capital_cost"],
-            marginal_cost=costs.at[tech, "marginal_cost"],
-            efficiency_store=costs.at[
-                lookup_store["battery"], "efficiency"
-            ],  # Lead_acid and lithium have the same value
-            efficiency_dispatch=costs.at[
-                lookup_dispatch["battery"], "efficiency"
-            ],  # Lead_acid and lithium have the same value
-            max_hours=max_hours["battery"],  # Lead_acid and lithium have the same value
-            cyclic_state_of_charge=True,
-        )
+    # Iterate through each microgrid
+    for microgrid in microgrid_ids:
+        # Iterate through each storage technology
+        for tech in technologies:
+            # Add the storage units to the power network
+            n.madd(
+                "StorageUnit",
+                {microgrid},
+                # buses_i
+                " " + tech,
+                bus=f"new_bus_{microgrid}",
+                carrier=tech,
+                p_nom_extendable=True,
+                capital_cost=costs.at[tech, "capital_cost"],
+                marginal_cost=costs.at[tech, "marginal_cost"],
+                efficiency_store=costs.at[
+                    lookup_store["battery"], "efficiency"
+                ],  # Lead_acid and lithium have the same value
+                efficiency_dispatch=costs.at[
+                    lookup_dispatch["battery"], "efficiency"
+                ],  # Lead_acid and lithium have the same value
+                max_hours=max_hours[
+                    "battery"
+                ],  # Lead_acid and lithium have the same value
+                cyclic_state_of_charge=True,
+            )
 
 
-def attach_load(n, load_file, tech_modelling):
+def attach_load(n, load_file, number_microgrids, tech_modelling):
+    # Upload the load csv file
     load = pd.read_csv(load_file).set_index([n.snapshots])
+    number_microgrids = len(number_microgrids.keys())
+    number_microgrids = range(0, number_microgrids)
 
-    # Number of loads
-    n_load = 1
-
-    # Create an index for the loads
-    index = pd.Index(list(range(n_load)))
-
-    # Get the index of the buses in the power network
-    buses_i = n.buses.index
+    # Rename the header of the demand df as the bus the load will be attached to
+    for i in number_microgrids:
+        load.rename(columns={f"{i+1}": f"new_bus_microgrid_{i+1}"}, inplace=True)
```

Comment:
This is something that I don't undestand; why is this needed?
My feeling is that the naming of the buses/micorgrids shall be more consistent and it may give problems along the chain.
The name of the bus of a microgrid shall be the name of the microgrid as specified in the config. I don't think there is the need to name them in the "hard" way inside the code.

Reply:
I needed that because I want to attach the load to the central buses of each grid (each one named "new_bus_microgrid_{i}). Anyway, I made some changes in this function and in build_demand so that the process of attaching the load is much more similar to pypsa earth